delfin/drivers/dell_emc/vmax/client.py

```diff
@@ -287,6 +287,118 @@ def list_volumes(self, storage_id):
             LOG.error(msg)
             raise exception.StorageBackendException(msg)
 
+    def list_controllers(self, storage_id):
+        try:
+            # Get list of Directors
+            directors = self.rest.get_director_list(self.array_id,
+                                                    self.uni_version)
+            controller_list = []
+            for director in directors:
+                director_info = self.rest.get_director(
+                    self.array_id, self.uni_version, director)
+
+                status = constants.ControllerStatus.NORMAL
+                if director_info.get('availability', '').upper() != 'ONLINE':
+                    status = constants.ControllerStatus.OFFLINE
+
+                controller = {
+                    'name': director_info['directorId'],
+                    'storage_id': storage_id,
+                    'native_controller_id': director_info['directorId'],
+                    'status': status,
+                    'location':
+                        'slot_' +
+                        str(director_info.get('director_slot_number')),
+                    'soft_version': None,
+                    'cpu_info': 'Cores-'
+                                + str(director_info.get('num_of_cores')),
+                    'memory_size': None
+
+                }
+                controller_list.append(controller)
+            return controller_list
+
+        except exception.SSLCertificateFailed:
+            LOG.error('SSL certificate failed when list pools for VMax')
+            raise
+        except Exception as err:
+            msg = "Failed to get controller metrics from VMAX: {}".format(err)
+            LOG.error(msg)
+            raise exception.ControllerNotFound(self.array_id)
+
+    def list_ports(self, storage_id):
+        try:
+            # Get list of Directors
+            directors = self.rest.get_director_list(self.array_id,
+                                                    self.uni_version)
+            switcher = {
+                'A': constants.PortLogicalType.MANAGEMENT,
```

Comment:
Why is the switcher keys (A,B,C,D) . It has to be (FA, RF, SE ) etc right?

Reply:
In VMAX director name, (Eg. `FA-1D`)  last letter is for Slice. Slice A is for Infrastructure Manager system emulation. Slice B is for HYPERMAX OS Data Services, Slice C for Backend emulation, Slice D-H remaining emulation.

Reply:
That means D-H slice can be anything  (Front End (FA), RDF( RF).. So I might see a port number RF-1D also.. I still believe the the first 2 letters should be used..   below are my understanding.. though I dont  see  any document explain this..
[DF,DA,DX]-> backend
[RF.RA,RE]-> RDF replication
[FA, FE, EA, EF, SE]-> Front End
[IM] - Infrastrucre manager

This doc has some info https://psrathore244.wordpress.com/2020/09/18/vmax-3-emulation/